src/functions-reference/higher-order_functions.Rmd

```diff
@@ -141,6 +142,131 @@ Stan's automatic differentiation.
+\index{{\tt \bfseries ode\_adams }!{\tt (function ode, vector initial\_state, real initial\_time, real[] times, ...): vector[]}|hyperpage}
+
+`vector[]` **`ode_adams`**`(function ode, vector initial_state, real initial_time, real[] times, ...)`<br>\newline
+Solves the ODE system for the times provided using the Adams-Moulton
+method with the implementation from CVODES.
+
+<!-- vector[]; ode_adams_tol; (function ode, vector initial_state, real initial_time, real[] times, data real rel_tol, data real abs_tol, data int max_num_steps, ...); -->
+\index{{\tt \bfseries ode\_adams\_tol }!{\tt (function ode, vector initial\_state, real initial\_time, real[] times, data real rel\_tol, data real abs\_tol, data int max\_num\_steps, ...): vector[]}|hyperpage}
+
+`vector[]` **`ode_adams_tol`**`(function ode, vector initial_state, real initial_time, real[] times, data real rel_tol, data real abs_tol, data int max_num_steps, ...)`<br>\newline
+Solves the ODE system for the times provided using the Adams-Moulton
+method with the implementation from CVODES with additional control
+parameters for the CVODES solver.
+
+### Stiff Solver
+
+<!-- vector[]; ode_bdf; (function ode, vector initial_state, real initial_time, real[] times, ...); -->
+\index{{\tt \bfseries ode\_bdf }!{\tt (function ode, vector initial\_state, real initial\_time, real[] times, ...): vector[]}|hyperpage}
+
+`vector[]` **`ode_bdf`**`(function ode, vector initial_state, real initial_time, real[] times, ...)`<br>\newline
+Solves the ODE system for the times provided using the backward
+differentiation formula (BDF) method with the implementation from
+CVODES.
+
+<!-- vector[]; ode_bdf_tol; (function ode, vector initial_state, real initial_time, real[] times, data real rel_tol, data real abs_tol, data int max_num_steps, ...); -->
+\index{{\tt \bfseries ode\_bdf\_tol }!{\tt (function ode, vector initial\_state, real initial\_time, real[] times, data real rel\_tol, data real abs\_tol, data int max\_num\_steps, ...): vector[]}|hyperpage}
+
+`vector[]` **`ode_bdf_tol`**`(function ode, vector initial_state, real initial_time, real[] times, data real rel_tol, data real abs_tol, data int max_num_steps, ...)`<br>\newline
+Solves the ODE system for the times provided using the backward
+differentiation formula (BDF) method with the implementation from
+CVODES with additional control parameters for the CVODES solver.
+
+### Arguments to the ODE solvers
+
+The arguments to the ODE solvers in both the stiff and non-stiff cases
+are as follows.
+
+*   *`ode`*: ODE system function with signature `(real, vector, ...):vector`
+
+The arguments represent (1) time, (2) system state, and a sequence of arguments of any type passed unmodified from the integration function call. The return value contains the derivatives with respect to time of the state.
+
+*   *`initial_state`*: initial state, type `vector`,
+
+*   *`initial_time`*: initial time, type `int` or `real`,
```

Comment:
We allow "int" as initial time... really? Is this cast into a real internally?

Reply:
Any function that can take a `real` argument should be able to take an `int` argument by casting it to `real`.  This does not (yet) apply to `real[]` vs. `int[]` arrays, I'm afraid, but should hold for primitives.

Reply:
This is one of those things that's for convenience as much as anything. Something like:

```
ode_bdf(f, y0, 0, ts, ...)
```

will cause an error if only real is supported because the 0 gets interpreted as an int instead of a double. int should promote to var or double, but I vaguely remember having this problem at one point anyway. Not sure the exact conditions.